pkg/docker/docker.go

```diff
@@ -0,0 +1,277 @@
+// package docker implements a docker launchable type
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"gopkg.in/v1/yaml"
+
+	"github.com/square/p2/pkg/cgroups"
+	"github.com/square/p2/pkg/launch"
+	"github.com/square/p2/pkg/runit"
+	"github.com/square/p2/pkg/util"
+	"github.com/square/p2/pkg/util/size"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	dockerclient "github.com/docker/docker/client"
+
+	// add this import explicitly for unix
+	_ "github.com/opencontainers/runc/libcontainer/system"
+)
+
+var _ launch.Launchable = &Launchable{}
+
+type Mounts map[string]Mount
+
+type Mount struct {
+	Type   string `yaml:"Type"`
+	Source string `yaml:"Source"`
+	Target string `yaml:"Target"`
+}
+
+type Launchable struct {
+	// LaunchableID is the ID of the launchable within the pod manifest, not to be
+	// confused with the container ID or the container name we'll use to
+	// identify the container with docker.
+	LaunchableID launch.LaunchableID
+
+	// The name of the docker image to run, as if passed to the docker
+	// command line
+	Image string
+
+	// RunAs is the name of the user the container should run as inside the container
+	RunAs string
+
+	// RootDir is where P2 writes files related to this container, not to be
+	// confused with the container filesystem's location on disk which is managed
+	// by dockerd
+	RootDir string
+
+	// RestartTimeout is poorly named in the context of docker containers, since
+	// we never restart a container, only stop them and start them. In this case
+	// it's the amount of time we ask docker to give a container to stop after
+	// SIGTERM is sent before SIGKILL
+	RestartTimeout time.Duration
+
+	// Unique name for log messages to uniquely identify the pod and container
+	ServiceID_ string
+
+	// DockerClient is the client used to start and stop containers
+	DockerClient *dockerclient.Client
+
+	// EntryPoints are the commands to run in the container
+	EntryPoints []string
+
+	// ParentCgroupID is the name of the pod-wide cgroup. Docker containers will
+	// be configured with cgroups inheriting from this one.
+	ParentCgroupID string
+
+	// CPUQuota is the number of CPUs the cgroup is allowed to use. Technically, this
+	// measurement is in "millions of microseconds per second", so it should be multiplied
+	// by a million before being passed to docker
+	CPUQuota int
+
+	// CgroupMemorySize is the max allowed memory docker should set for the cgroup for
+	// the container
+	CgroupMemorySize size.ByteCount
+
+	// RestartPolicy_ determines what behavior docker should have when the pod exits.
+	// Typically it will indicate that docker should always restart the container
+	// for long-running services (e.g. servers)
+	RestartPolicy_ runit.RestartPolicy
+}
+
+func (l *Launchable) Disable() error {
```

Comment:
not related to this PR, but maybe we do want this actually, e.g. kubernetes allows containers to specify a PreStop hook: https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/

Reply:
I have a use case where this might be nice